src/data_readers/data_reader_hdf5.cpp

```diff
@@ -0,0 +1,237 @@
+        }
+
+        // load in dataset
+        hid_t h_data =  H5Dopen(h_file, HDF5_KEY_DATA.c_str(), H5P_DEFAULT);
+        hid_t filespace = H5Dget_space(h_data);
+        int rank1 = H5Sget_simple_extent_ndims(filespace);
+        hsize_t dims[rank1];
+        H5Sget_simple_extent_dims(filespace, dims, NULL);
+            
+        if (h_data < 0) {
+            throw lbann_exception(std::string{} + __FILE__ + " " + std::to_string(__LINE__) +
+                                        " hdf5_reader::load() - can't find hdf5 key : " + HDF5_KEY_DATA);
+        }
+
+        //TODO: add the int 16 stuff
+	    //TODO: change the indexing 
+	    // check if mb_idx needs to be changed to not be hard coded
+	    int adj_mb_idx = mb_idx+(rank%4);
+        Mat X_v = El::View(X, El::IR(0,X.Height()), El::IR(adj_mb_idx, adj_mb_idx+1));
+
+        DataType *dest = X_v.Buffer();
+        //TODO Add a reference to X mat so only one read
+	    //will this work ? ?
+        read_hdf5(h_data, filespace, world_rank, HDF5_KEY_DATA, dims, dest);
+        //close data set
+        H5Dclose(h_data);
+        if (m_has_responses) {
+            h_data = H5Dopen(h_file, HDF5_KEY_RESPONSES.c_str(), H5P_DEFAULT);
+            H5Dread(h_data, H5T_NATIVE_DOUBLE, H5S_ALL, H5S_ALL, H5P_DEFAULT, m_all_responses);
+            H5Dclose(h_data);
+        }
+        H5Fclose(h_file);
+        double end_file= MPI_Wtime();
+        std::cerr<<"per file time " << end_file -start_file << "seconds to run. \n";
+       }
+
+        //TODO do i need this?
+	// not if I pass a ref to X I dont think
+	//this should be equal to num_nuerons/LBANN_NUM_IO_PARTITIONS
+        //unsigned long int pixelcount = m_image_width*m_image_height*m_image_depth*m_image_num_channels;
+  // #ifdef LBANN_DISTCONV_COSMOFLOW_KEEP_INT16
+    //    std::memcpy(dest,data, sizeof(short)*pixelcount);
+  // #else
+    //    LBANN_OMP_PARALLEL_FOR
+     //       for(int p = 0; p<pixelcount; p++) {
+                //TODO what is m_scaling_factor_int16
+     //           dest[p] = tmp[p] * m_scaling_factor_int16;
+                // mash this with above
+                    //X.Set(p, mb_idx,*tmp++);
+     //       }
+  // #endif
+        //auto pixel_col = X(El::IR(0, X.Height()), El::IR(mb_idx, mb_idx+1));
+        //std::vector<size_t> dims = {
+          //  1ull,
+          //  static_cast<size_t>(m_image_height),
+          //  static_cast<size_t>(m_image_width)};
+        //m_transform_pipeline.apply(pixel_col, dims);
+        prof_region_end("fetch_datum", false);
+        return true;
+    }
+    //get from a cached response
+    bool hdf5_reader::fetch_response(Mat& Y, int data_id, int mb_idx) {
+        prof_region_begin("fetch_response", prof_colors[0], false);
+        Mat Y_v = El::View(Y, El::IR(0, Y.Height()), El::IR(mb_idx, mb_idx+1));
+        //TODO: possibly 4 tho, python tells me its float64
+        std::memcpy(Y_v.Buffer(), &m_all_responses,
```

Comment:
Where is `m_all_responses` defined? 

I believe it should exist in `hdf5_reader` class. Make it `float`, not `double`.